### PR TITLE
Change parse_infix_expression to use combinator style.

### PR DESCRIPTION
**Pull request by gittywithexcitement**
- Change parse_infix_expression to use combinator style.
- Change parse_infix_expression and parse_expression to take 1 arg and return a parser.
- Make some things Clone so that the above works.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -295,70 +295,81 @@ fn parse_unary_atomic_expression(input: ParseInput) -> ParseResult<Expression> {
...
-    Ok((
-        rest,
-        Spanned {
-            span: Span::new(start, end),
-            thing: re,
-        },
-    ))
+    // Note: the spanned combinator can't be used here because `left` was passed to us (thus,
+    // spanned can't capture the start).
+    move |input: ParseInput<'a>| -> ParseResult<Expression> {
```
Comment (RobSpringer) [inline src/lib.rs:310]: While this is more "elegant", it sure is harder to read for a newbie :\

But this is how we learn; I suspect everything here is really standard Rust. It's just newish to me.

Comment (RobSpringer) [reply src/lib.rs:310]: But a map over a map is...a brainful.

(BRB gotta look up flat_map vs map)

Comment (RobSpringer) [reply src/lib.rs:310]: Ok. After reading this, this is a lot less complicated/dramatic a change as I initially feared. LGTM!

**Review by RobSpringer — APPROVED** [review]
LGTM! Just some train-of-thought comments. I could delete them, but w/e.

**Comment by gittywithexcitement** [comment]
> LGTM! Just some train-of-thought comments. I could delete them, but w/e.

Haha, it's cool, I operate the same way. :)

**Pull request merged**